2021/day16/day16.java

```diff
@@ -0,0 +1,151 @@
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Scanner;
+
+public class day16 {
+    public static class Packet {
+        Long value;
+        int length;
+        int versionSum;
+        String binData;
+    }
+
+    public static String readInput(String filename) {
+        String s = "";
+        String result = "";
+        Scanner sc = null;
+        try {
+            sc = new Scanner(new File(filename));
+            s = sc.next();
+            for (char c : s.toCharArray()) {
+                result += hexToBin.get(String.valueOf(c));
+            }
+        } catch (Exception e) {
+            System.out.println("Something went wrong " + e);
+        } finally {
+            sc.close();
+        }
+        return result;
+    }
+
+    private static final HashMap<String, String> hexToBin = new HashMap<>();
+    static {
+        hexToBin.put("0", "0000");
+        hexToBin.put("1", "0001");
+        hexToBin.put("2", "0010");
+        hexToBin.put("3", "0011");
+        hexToBin.put("4", "0100");
+        hexToBin.put("5", "0101");
+        hexToBin.put("6", "0110");
+        hexToBin.put("7", "0111");
+        hexToBin.put("8", "1000");
+        hexToBin.put("9", "1001");
+        hexToBin.put("A", "1010");
+        hexToBin.put("B", "1011");
+        hexToBin.put("C", "1100");
+        hexToBin.put("D", "1101");
+        hexToBin.put("E", "1110");
+        hexToBin.put("F", "1111");
+    }
```

Comment:
This can be made easier by using built-in Java functionality - try to implement it yourself. As a help, I can suggest using `String.format`, `Integer.toBinaryString` and `Integer.parseInt(_, 16)`